### PR TITLE
native: support for compile-time conditionals

### PR DESCRIPTION
**Pull request by Spydr06**
This PR adds support for some compile-time conditionals in the native backend.

### CODE REVIEW

**Comment by Spydr06** [comment]
Also, I tried implementing `AtExpr`, but the `.val` field seems to always be an empty string

File: vlib/v/gen/native/tests/comptime.vv.out [hunk 1]
```diff
@@ -0,0 +1,3 @@
+linux or windows or macos
+custom defines work
+other
```
Comment (Delta456) [inline vlib/v/gen/native/tests/comptime.vv.out:3]: Add a new line here

Comment (Spydr06) [reply vlib/v/gen/native/tests/comptime.vv.out]: done

**Comment by spytheman** [comment]
> Also, I tried implementing `AtExpr`, but the `.val` field seems to always be an empty string

Try with:
```v
fn abc() {
        // a comment
        a := @LINE
        b := @FILE
        c := @FN
        println(a)
        println(b)
        println(c)
}

fn main() {
        abc()
}
```

**Comment by Spydr06** [comment]
In this context, it seems to work, but something like
```v
println(@LINE)
```
doesn't work, because the `.val` field is empty

**Comment by spytheman** [comment]
> ```v
> println(@LINE)
> ```

I see ... Well, that can be filed as a separate issue.

I think it can be merged as it is, since it is still much better than it was before, and that problem can be solved in another PR.

**Comment by Spydr06** [comment]
sounds good

**Comment by spytheman** [comment]
It seems that `println(anything_complex)` does not currently work, because:
```
vlib/v/checker/fn.v-690-if !found && c.pref.backend == .native {
vlib/v/checker/fn.v:715:if is_native_builtin {
vlib/v/checker/fn.v-716-return ast.void_type
vlib/v/checker/fn.v-717-}
```
i.e. the Checker.fn_call method returns early, without checking any of the parameters for `println()` when the native backend is chosen => the expressions in the AST for the params are not checked => in this particular case, ast.AtExpr.val is not evaluated .

**Comment by Spydr06** [comment]
Should I fix this, or is this something for the next PR?

**Comment by spytheman** [comment]
A separate PR will be better. I think it was done that way, so that `println` could be bootstrapped without having to read `builtin` and generate code for it. Fixing that is out of scope for this PR.

**Comment by Spydr06** [comment]
ok, thank you

**Review by spytheman — APPROVED** [review]
Excellent work.

**Comment by Spydr06** [comment]
thank you very much! :)

**Comment by Spydr06** [comment]
> A separate PR will be better. I think it was done that way, so that `println` could be bootstrapped without having to read `builtin` and generate code for it. Fixing that is out of scope for this PR.

Well, I just fixed it. Should I push the changes?

**Comment by Spydr06** [comment]
Here are the changes:
```diff
diff --git a/vlib/v/checker/fn.v b/vlib/v/checker/fn.v
index c42239c30..e84662df9 100644
--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -487,6 +487,38 @@ pub fn (mut c Checker) call_expr(mut node ast.CallExpr) ast.Type {
 	return typ
 }
 
+pub fn (mut c Checker) builtin_args(mut node ast.CallExpr, fn_name string, func ast.Fn) {
+	c.inside_println_arg = true
+	c.expected_type = ast.string_type
+	node.args[0].typ = c.expr(node.args[0].expr)
+	arg := node.args[0]
+	c.check_expr_opt_call(arg.expr, arg.typ)
+	if arg.typ.is_void() {
+		c.error('`$fn_name` can not print void expressions', node.pos)
+	} else if arg.typ == ast.char_type && arg.typ.nr_muls() == 0 {
+		c.error('`$fn_name` cannot print type `char` directly, print its address or cast it to an integer instead',
+			node.pos)
+	}
+	c.fail_if_unreadable(arg.expr, arg.typ, 'argument to print')
+	c.inside_println_arg = false
+	node.return_type = ast.void_type
+	c.set_node_expected_arg_types(mut node, func)
+
+	/*
+	// TODO: optimize `struct T{} fn (t &T) str() string {return 'abc'} mut a := []&T{} a << &T{} println(a[0])`
+	// It currently generates:
+	// `println(T_str_no_ptr(*(*(T**)array_get(a, 0))));`
+	// ... which works, but could be just:
+	// `println(T_str(*(T**)array_get(a, 0)));`
+	prexpr := node.args[0].expr
+	prtyp := node.args[0].typ
+	prtyp_sym := c.table.sym(prtyp)
+	prtyp_is_ptr := prtyp.is_ptr()
+	prhas_str, prexpects_ptr, prnr_args := prtyp_sym.str_method_info()
+	eprintln('>>> println hack typ: ${prtyp} | sym.name: ${prtyp_sym.name} | is_ptr: $prtyp_is_ptr | has_str: $prhas_str | expects_ptr: $prexpects_ptr | nr_args: $prnr_args | expr: ${prexpr.str()} ')
+	*/
+}
+
 pub fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool) ast.Type {
 	fn_name := node.name
 	if fn_name == 'main' {
@@ -713,6 +745,9 @@ pub fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool)
 		}
 	}
 	if is_native_builtin {
+		if node.args.len > 0 && fn_name in ['println', 'print', 'eprintln', 'eprint', 'panic'] {
+			c.builtin_args(mut node, fn_name, func)
+		}
 		return ast.void_type
 	}
 	// check for arg (var) of fn type
@@ -841,34 +876,7 @@ pub fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool)
 	}
 	// println / eprintln / panic can print anything
 	if node.args.len > 0 && fn_name in ['println', 'print', 'eprintln', 'eprint', 'panic'] {
-		c.inside_println_arg = true
-		c.expected_type = ast.string_type
-		node.args[0].typ = c.expr(node.args[0].expr)
-		arg := node.args[0]
-		c.check_expr_opt_call(arg.expr, arg.typ)
-		if arg.typ.is_void() {
-			c.error('`$fn_name` can not print void expressions', node.pos)
-		} else if arg.typ == ast.char_type && arg.typ.nr_muls() == 0 {
-			c.error('`$fn_name` cannot print type `char` directly, print its address or cast it to an integer instead',
-				node.pos)
-		}
-		c.fail_if_unreadable(arg.expr, arg.typ, 'argument to print')
-		c.inside_println_arg = false
-		node.return_type = ast.void_type
-		c.set_node_expected_arg_types(mut node, func)
-		/*
-		// TODO: optimize `struct T{} fn (t &T) str() string {return 'abc'} mut a := []&T{} a << &T{} println(a[0])`
-		// It currently generates:
-		// `println(T_str_no_ptr(*(*(T**)array_get(a, 0))));`
-		// ... which works, but could be just:
-		// `println(T_str(*(T**)array_get(a, 0)));`
-		prexpr := node.args[0].expr
-		prtyp := node.args[0].typ
-		prtyp_sym := c.table.sym(prtyp)
-		prtyp_is_ptr := prtyp.is_ptr()
-		prhas_str, prexpects_ptr, prnr_args := prtyp_sym.str_method_info()
-		eprintln('>>> println hack typ: ${prtyp} | sym.name: ${prtyp_sym.name} | is_ptr: $prtyp_is_ptr | has_str: $prhas_str | expects_ptr: $prexpects_ptr | nr_args: $prnr_args | expr: ${prexpr.str()} ')
-		*/
+		c.builtin_args(mut node, fn_name, func)
 		return func.return_type
 	}
 	// `return error(err)` -> `return err`
```

**Comment by spytheman** [comment]
> Well, I just fixed it. Should I push the changes?

Please make a separate PR for it.

**Comment by Spydr06** [comment]
Ok good, thank you

**Pull request merged**